modules/services/hyprpaper.nix

```
@@ -0,0 +1,90 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+
+  cfg = config.services.hyprpaper;
+in {
+  meta.maintainers = [ maintainers.khaneliman maintainers.fufexan ];
+
+  options.services.hyprpaper = {
+    enable = mkEnableOption "Hyprpaper, Hyprland's wallpaper daemon";
+
+    package = mkPackageOption pkgs "hyprpaper" { };
+
+    settings = lib.mkOption {
+      type = with lib.types;
+        let
+          valueType = nullOr (oneOf [
+            bool
+            int
+            float
+            str
+            path
+            (attrsOf valueType)
+            (listOf valueType)
+          ]) // {
+            description = "Hyprpaper configuration value";
+          };
+        in valueType;
+      default = { };
+      description = ''
+        hyprpaper configuration written in Nix. Entries with the same key
+        should be written as lists. Variables' and colors' names should be
+        quoted. See <https://wiki.hyprland.org> for more examples.
+      '';
+      example = lib.literalExpression ''
+        {
+          general = {
+            after_sleep_cmd = "hyprctl dispatch dpms on";
+            ignore_dbus_inhibit = false;
+            lock_cmd = "hyprlock";
+          };
+
+          listener = [
+            {
+              timeout = 900;
+              on-timeout = "hyprlock";
+            }
+            {
+              timeout = 1200;
+              on-timeout = "hyprctl dispatch dpms off";
+              on-resume = "hyprctl dispatch dpms on";
+            }
+          ];
+        }
+      '';
+    };
+
+    importantPrefixes = lib.mkOption {
+      type = with lib.types; listOf str;
+      default = [ "$" ];
+      example = [ "$" ];
+      description = ''
+        List of prefix of attributes to source at the top of the config.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    xdg.configFile."hypr/hyprpaper.conf" = mkIf (cfg.settings != { }) {
+      text = lib.hm.generators.toHyprconf {
+        attrs = cfg.settings;
+        inherit (cfg) importantPrefixes;
+      };
+    };
+
+    systemd.user.services.hyprpaper = {
+      Unit = {
+        Description = "hyprpaper";
+        After = [ "graphical-session.target" ];
+        WantedBy = [ "default.target" ];
```

Shouldn't this be the more conventional

```suggestion
        After = [ "graphical-session-pre.target" ];
        PartOf = [ "graphical-session.target" ];
```

?

Also, need to add

```
X-Restart-Triggers = [ "${config.xdg.configFile."hypr/hyprpaper.conf".source}" ];
```

to enable automatic restart when the configuration changes.

---

Also, you'll probably need

``` nix
Install = { WantedBy = [ "graphical-session.target" ]; };
```

to get it started.


And

``nix
ConditionEnvironment = [ "WAYLAND_DISPLAY" ];
```

to only start for Wayland sessions…

---

Btw, an alternative to `X-Restart-Triggers` would be to provide the path to the configuration file directly in the command. Something like

```nix
ExecStart = let
  configFile = pkgs.writeText "hyprpaper.conf"
    (lib.hm.generators.toHyprconf {
      attrs = cfg.settings;
      inherit (cfg) importantPrefixes;
    });
in "${getExe cfg.package} -c '${configFile}'";
```

perhaps.

---

I'll implement with the X-restart-triggers since it seems more flexible with supporting people who just want the service and do the config file outside the module, or even wanting to test a different binary or test config file not through the service since the program will read from the expected file location.

---

Did the same update to hypridle's PR.